src/bsoncxx/types/value.cpp

```diff
@@ -40,15 +45,32 @@ value::value(const value& rhs) {
 }
 
 value& value::operator=(const value& rhs) {
-    _type = rhs._type;
+    if (this == &rhs) {
+        return *this;
+    }
+
+    if (_type) {
+            switch (static_cast<typename std::underlying_type<bsoncxx::type>::type>(*_type)) {
+#define BSONCXX_ENUM(type, val)     \
+        case val:                   \
+            _b_##type.~b_##type();  \
```

Comment:
This implementation strategy means that value doesn't offer the strong exception safety guarantee.